### DISCUSSION TITLE
Codegen and different component props between iOS and Android

### DISCUSSION THREAD

**Discussion by gispada** [reactwg/react-native-new-architecture#76] [Q&A] [answered]
When defining Codegen specifications for a Fabric component, what's the best way to deal with different props between platforms?

It seems that every file ending with _NativeComponent_ gets picked out by Codegen, am I right? So it might generate code for props that will never be used on both platforms.

Setting `jsRootDir` in android/build.gradle to a different folder than `jsSrcsDir` in package.json's `codegenConfig` can be a valid approach?

**Comment by gispada** [comment]
Upon further investigation, I found out that it's possible to specify which platform to exclude directly in the _codegenNativeComponent_ function:
```ts
export default codegenNativeComponent<NativeProps>('RNComponent', {
  excludedPlatforms: ['android'],
}) as HostComponent<NativeProps>;
```
Digging into Codegen source code it also seems that — for modules only — it's possible to separate platforms by appending `IOS`, `Android` or `Cxx` to the module filename.

One thing that's not entirely clear is why C++ files in `android/build/generated/source/codegen/jni/react/renderer/components/MyComponent` contain references to those components that should be excluded. They are indeed excluded from the java generated code (no interface/delegate), but are still appearing in the jni layer.

**Comment by cortinico** [comment, marked as answer]
Thanks for your investigation @gispada 
This is currently not easily doable out of the box. The current solution is either to:
- Have a spec which is a superset of the capabilities of the two platforms, and stub the implementation for platform specific APIs
- Use `excludedPlatforms: ['...']` as you suggest. You'll be effectively not generating spec for one of the two platform.
- Use the module filename suffixes as you suggest. This is however not documented and we're probably going to revist this.

Ideally, the codegen should be able to handle `.android.js` and `.ios.js` file extensions.

> One thing that's not entirely clear is why C++ files in android/build/generated/source/codegen/jni/react/renderer/components/MyComponent contain references to those components that should be excluded

This is a bug to me and should be fixed. Not a major one as you will probably ignore those sources, but still somethign to clean up.
